File: revos-server/app/utils/plaid.py
"""
Plaid Integration Service
Handles bank account linking, transaction retrieval, and spending analytics
"""

import os
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
import plaid
from plaid.api import plaid_api
from plaid.model.transactions_get_request import TransactionsGetRequest
from plaid.model.accounts_get_request import AccountsGetRequest
from plaid.model.item_public_token_exchange_request import ItemPublicTokenExchangeRequest
from plaid.model.transactions_recurring_get_request import TransactionsRecurringGetRequest

logger = logging.getLogger(__name__)

# Initialize Plaid client with explicit credentials from environment
# The SDK needs these passed explicitly, not auto-loaded
client_id = os.getenv('PLAID_CLIENT_ID')
secret = os.getenv('PLAID_SECRET')
api_key = os.getenv('PLAID_API_KEY')

if not client_id or not secret:
    logger.warning("⚠️  Plaid credentials not found in environment variables!")
    logger.warning("   Make sure PLAID_CLIENT_ID and PLAID_SECRET are set in .env")

configuration = plaid.Configuration(
    host=plaid.Environment.Sandbox,  # Use Production for live
)

# Manually set credentials on the configuration
configuration.api_key['clientId'] = client_id
configuration.api_key['secret'] = secret
configuration.api_key['plaidVersion'] = '2020-09-14'

api_client = plaid.ApiClient(configuration)
client = plaid_api.PlaidApi(api_client)


class PlaidManager:
    """Manages Plaid API interactions for bank account integration"""

    @staticmethod
    def create_link_token(user_id: str, client_name: str = "RevOS") -> Dict[str, Any]:
        """
        Create a Plaid Link token for initiating account connection
        
        Args:
            user_id: User identifier for session management
            client_name: Name of the client (for Plaid Link)
            
        Returns:
            Dictionary containing link_token and expiration
        """
        try:
            from plaid.model.link_token_create_request import LinkTokenCreateRequest
            from plaid.model.link_token_create_request_user import LinkTokenCreateRequestUser
            from plaid.model.country_code import CountryCode
            from plaid.model.products import Products

            request = LinkTokenCreateRequest(
                products=[Products("auth"), Products("transactions")],
                client_name=client_name,
                country_codes=[CountryCode("US")],
                language="en",
                user=LinkTokenCreateRequestUser(
                    client_user_id=str(user_id),
                ),
            )

            response = client.link_token_create(request)
            logger.info(f"✅ Created Plaid Link token for user {user_id}")
            
            return {
                "link_token": response.link_token,
                "expiration": response.expiration.isoformat() if response.expiration else None,
            }
        except Exception as e:
            logger.error(f"❌ Error creating Plaid Link token: {e}")
            raise

    @staticmethod
    def exchange_public_token(public_token: str, metadata: Dict[str, Any]) -> Dict[str, Any]:
        """
        Exchange Plaid public token for access token
        
        Args:
            public_token: Public token from Plaid Link
            metadata: Metadata about connected institution and accounts
            
        Returns:
            Dictionary containing access_token and item_id
        """
        try:
            request = ItemPublicTokenExchangeRequest(public_token=public_token)
            response = client.item_public_token_exchange(request)

            logger.info(f"✅ Exchanged public token for item {response.item_id}")
            
            return {
                "access_token": response.access_token,
                "item_id": response.item_id,
                "institution": metadata.get("institution", {}),
                "accounts": metadata.get("accounts", []),
            }
        except Exception as e:
            logger.error(f"❌ Error exchanging public token: {e}")
            raise

    @staticmethod
    def get_accounts(access_token: str) -> List[Dict[str, Any]]:
        """
        Get list of accounts for an access token
        
        Args:
            access_token: Plaid access token
            
        Returns:
            List of account information
        """
        try:
            request = AccountsGetRequest(access_token=access_token)
            response = client.accounts_get(request)

            accounts = []
            for account in response.accounts:
                accounts.append({
                    "account_id": account.account_id,
                    "name": account.name,
                    "type": account.type,
                    "subtype": account.subtype,
                    "mask": account.mask,
                    "balance": {
                        "available": account.balances.available,
                        "current": account.balances.current,
                        "limit": account.balances.limit,
                    }
                })

            logger.info(f"✅ Retrieved {len(accounts)} accounts")
            return accounts
        except Exception as e:
            logger.error(f"❌ Error getting accounts: {e}")
            raise

    @staticmethod
    def get_transactions(
        access_token: str,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get transactions for an access token
        
        Args:
            access_token: Plaid access token
            start_date: Start date (YYYY-MM-DD), defaults to 30 days ago
            end_date: End date (YYYY-MM-DD), defaults to today
            limit: Maximum number of transactions to retrieve
            
        Returns:
            List of transactions
        """
        try:
            # Default to last 30 days
            if not end_date:
                end_date = datetime.now().strftime("%Y-%m-%d")
            if not start_date:
                start_date = (datetime.now() - timedelta(days=30)).strftime("%Y-%m-%d")

            request = TransactionsGetRequest(
                access_token=access_token,
                start_date=start_date,
                end_date=end_date,
                options={"count": limit, "offset": 0},
            )

            response = client.transactions_get(request)

            transactions = []
            for transaction in response.transactions:
                transactions.append({
                    "transaction_id": transaction.transaction_id,
                    "name": transaction.name,
                    "amount": transaction.amount,
                    "date": transaction.date.isoformat(),
                    "category": transaction.category,
                    "merchant_name": transaction.merchant_name,
                    "account_id": transaction.account_id,
                    "pending": transaction.pending,
                })

            logger.info(f"✅ Retrieved {len(transactions)} transactions from {start_date} to {end_date}")
            return transactions
        except Exception as e:
            logger.error(f"❌ Error getting transactions: {e}")
            raise

    @staticmethod
    def get_spending_by_category(transactions: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Analyze spending by category
        
        Args:
            transactions: List of transactions
            
        Returns:
            Dictionary with spending analysis by category
        """
        try:
            spending_by_category = {}
            
            for transaction in transactions:
                category = transaction.get("category", ["Other"])[0] if transaction.get("category") else "Other"
                amount = transaction.get("amount", 0)
                
                if category not in spending_by_category:
                    spending_by_category[category] = {
                        "total": 0,
                        "count": 0,
                        "transactions": []
                    }
                
                spending_by_category[category]["total"] += amount
                spending_by_category[category]["count"] += 1
                spending_by_category[category]["transactions"].append({
                    "name": transaction.get("name"),
                    "amount": amount,
                    "date": transaction.get("date"),
                })

            # Sort by total spending
            sorted_categories = sorted(
                spending_by_category.items(),
                key=lambda x: x[1]["total"],
                reverse=True
            )

            logger.info(f"✅ Analyzed spending across {len(sorted_categories)} categories")
            
            return {
                "by_category": dict(sorted_categories),
                "total_spending": sum(cat["total"] for cat in spending_by_category.values()),
                "average_transaction": sum(cat["total"] for cat in spending_by_category.values()) / sum(cat["count"] for cat in spending_by_category.values()) if spending_by_category else 0,
            }
        except Exception as e:
            logger.error(f"❌ Error analyzing spending: {e}")
            raise

    @staticmethod
    def get_recurring_transactions(access_token: str) -> List[Dict[str, Any]]:
        """
        Get recurring transactions
        
        Args:
            access_token: Plaid access token
            
        Returns:
            List of recurring transactions
        """
        try:
            request = TransactionsRecurringGetRequest(access_token=access_token)
            response = client.transactions_recurring_get(request)

            recurring = []
            for transaction in response.recurring_transactions:
                recurring.append({
                    "name": transaction.merchant_name or transaction.description,
                    "amount": transaction.amounts[0]["amount"] if transaction.amounts else 0,
                    "frequency": transaction.frequency,
                    "last_date": transaction.last_amount_at.isoformat() if transaction.last_amount_at else None,
                    "next_expected_date": transaction.next_scheduled_transaction_date.isoformat() if transaction.next_scheduled_transaction_date else None,
                })

            logger.info(f"✅ Retrieved {len(recurring)} recurring transactions")
            return recurring
        except Exception as e:
            logger.error(f"❌ Error getting recurring transactions: {e}")
            # Recurring transactions may not be available in all environments
            logger.warning("Recurring transactions not available in this environment")
            return []

    @staticmethod
    def get_insights(access_token: str) -> Dict[str, Any]:
        """
        Generate spending insights and recommendations
        
        Args:
            access_token: Plaid access token
            
        Returns:
            Dictionary with spending insights
        """
        try:
            # Get transactions from last 90 days
            end_date = datetime.now().strftime("%Y-%m-%d")
            start_date = (datetime.now() - timedelta(days=90)).strftime("%Y-%m-%d")

            transactions = PlaidManager.get_transactions(access_token, start_date, end_date, 500)
            spending_analysis = PlaidManager.get_spending_by_category(transactions)

            # Calculate insights
            total_spending = spending_analysis["total_spending"]
            daily_average = total_spending / 90
            
            # Get top spending categories
            top_categories = sorted(
                spending_analysis["by_category"].items(),
                key=lambda x: x[1]["total"],
                reverse=True
            )[:5]

            insights = {
                "period": {
                    "start_date": start_date,
                    "end_date": end_date,
                    "days": 90,
                },
                "summary": {
                    "total_spending": total_spending,
                    "daily_average": daily_average,
                    "monthly_average": total_spending / 3,  # Approximate monthly
                    "transaction_count": sum(cat["count"] for cat in spending_analysis["by_category"].values()),
                },
                "top_categories": [
                    {
                        "category": cat[0],
                        "amount": cat[1]["total"],
                        "percentage": (cat[1]["total"] / total_spending * 100) if total_spending > 0 else 0,
                        "transaction_count": cat[1]["count"],
                    }
                    for cat in top_categories
                ],
                "recommendations": PlaidManager._generate_recommendations(spending_analysis),
            }

            logger.info(f"✅ Generated spending insights")
            return insights
        except Exception as e:
            logger.error(f"❌ Error generating insights: {e}")
            raise

    @staticmethod
    def _generate_recommendations(spending_analysis: Dict[str, Any]) -> List[str]:
        """
        Generate personalized spending recommendations
        
        Args:
            spending_analysis: Spending analysis data
            
        Returns:
            List of recommendations
        """
        recommendations = []
        
        try:
            categories = spending_analysis.get("by_category", {})
            
            # Check for high dining spending
            if "FOOD_AND_DRINK" in categories:
                amount = categories["FOOD_AND_DRINK"]["total"]
                if amount > 200:
                    recommendations.append("💰 Consider reducing dining out expenses - you're spending over $200/month here!")

            # Check for high shopping
            if "SHOPPING" in categories:
                amount = categories["SHOPPING"]["total"]
                if amount > 150:
                    recommendations.append("🛍️ Your shopping spending is high. Try setting a monthly budget!")

            # Check for subscriptions
            subscription_keywords = ["subscription", "spotify", "netflix", "hulu", "gym", "membership"]
            monthly_subscriptions = sum(
                cat["total"] for cat_name, cat in categories.items()
                if any(keyword in cat_name.lower() for keyword in subscription_keywords)
            )
            if monthly_subscriptions > 50:
                recommendations.append(f"📺 You're spending ${monthly_subscriptions:.2f}/month on subscriptions. Audit which ones you actually use!")

            # General recommendation if none triggered
            if not recommendations:
                recommendations.append("✅ Your spending looks balanced! Keep monitoring your budget.")

        except Exception as e:
            logger.error(f"Error generating recommendations: {e}")

        return recommendations

File: revos-client/src/services/plaid.ts
/**
 * Frontend Plaid Integration Service
 * Handles all Plaid functionality directly in React
 * No backend Plaid endpoints needed
 */

// Types
export interface PlaidAccount {
  id: string;
  name: string;
  type: string;
  subtype: string;
  mask: string;
  balance?: {
    available: number;
    current: number;
    limit?: number;
  };
}

export interface Transaction {
  transaction_id: string;
  name: string;
  amount: number;
  date: string;
  category: string[];
  merchant_name?: string;
  pending: boolean;
}

export interface SpendingInsights {
  period: {
    start_date: string;
    end_date: string;
    days: number;
  };
  summary: {
    total_spending: number;
    daily_average: number;
    monthly_average: number;
    transaction_count: number;
  };
  top_categories: Array<{
    category: string;
    amount: number;
    percentage: number;
    transaction_count: number;
  }>;
  recommendations: string[];
}

export interface LinkedAccount {
  id: string;
  institution_name: string;
  institution_id: string;
  accounts: PlaidAccount[];
  public_token?: string;
  access_token: string;
}

/**
 * PlaidService - All Plaid operations in React
 */
export class PlaidService {
  private static readonly STORAGE_KEY = 'plaid_linked_accounts';
  private static readonly TRANSACTIONS_KEY = 'plaid_transactions';
  private static readonly INSIGHTS_KEY = 'plaid_spending_insights';

  /**
   * Get Plaid Link token from backend
   * This is the only backend call needed
   */
  static async createLinkToken(userId: string, authToken: string): Promise<string> {
    try {
      const response = await fetch('http://localhost:5000/api/plaid/create-link-token', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${authToken}`,
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error('Failed to create link token');
      }

      const data = await response.json();
      return data.link_token;
    } catch (error) {
      console.error('❌ Error creating link token:', error);
      throw error;
    }
  }

  /**
   * Save linked account to localStorage
   * In production, you'd send this to backend to store encrypted
   */
  static saveLinkedAccount(account: LinkedAccount): void {
    try {
      const accounts = this.getLinkedAccounts();
      const existingIndex = accounts.findIndex(a => a.id === account.id);
      
      if (existingIndex >= 0) {
        accounts[existingIndex] = account;
      } else {
        accounts.push(account);
      }

      localStorage.setItem(this.STORAGE_KEY, JSON.stringify(accounts));
      console.log('✅ Account saved to localStorage');
    } catch (error) {
      console.error('❌ Error saving account:', error);
      throw error;
    }
  }

  /**
   * Get all linked accounts from localStorage
   */
  static getLinkedAccounts(): LinkedAccount[] {
    try {
      const data = localStorage.getItem(this.STORAGE_KEY);
      return data ? JSON.parse(data) : [];
    } catch (error) {
      console.error('❌ Error retrieving accounts:', error);
      return [];
    }
  }

  /**
   * Remove a linked account
   */
  static removeLinkedAccount(accountId: string): void {
    try {
      const accounts = this.getLinkedAccounts();
      const filtered = accounts.filter(a => a.id !== accountId);
      localStorage.setItem(this.STORAGE_KEY, JSON.stringify(filtered));
      console.log('✅ Account removed');
    } catch (error) {
      console.error('❌ Error removing account:', error);
      throw error;
    }
  }

  /**
   * Get transactions from Plaid API via backend
   * Fetches real transaction data for the selected account
   */
  static async getTransactions(
    accountId: string,
    authToken?: string,
    apiBaseUrl: string = 'http://localhost:5000'
  ): Promise<Transaction[]> {
    try {
      if (!authToken) {
        console.warn('⚠️ No auth token provided, returning empty transactions');
        return [];
      }

      // Get the account to retrieve its access_token
      const accounts = this.getLinkedAccounts();
      const account = accounts.find(a => a.id === accountId);
      
      if (!account || !account.access_token) {
        console.warn('⚠️ Account or access_token not found');
        return [];
      }

      const startDate = new Date();
      startDate.setDate(startDate.getDate() - 90); // Last 90 days
      const endDate = new Date();

      const response = await fetch(
        `${apiBaseUrl}/api/plaid/transactions?start_date=${startDate.toISOString().split('T')[0]}&end_date=${endDate.toISOString().split('T')[0]}&limit=100&access_token=${encodeURIComponent(account.access_token)}`,
        {
          method: 'GET',
          headers: {
            'Authorization': `Bearer ${authToken}`,
            'Content-Type': 'application/json',
          },
        }
      );

      if (!response.ok) {
        throw new Error(`Failed to fetch transactions: ${response.statusText}`);
      }

      const data = await response.json();
      const transactions = data.transactions || [];

      localStorage.setItem(this.TRANSACTIONS_KEY, JSON.stringify(transactions));
      return transactions;
    } catch (error) {
      console.error('❌ Error getting transactions:', error);
      return [];
    }
  }

  /**
   * Calculate spending insights from transactions
   */
  static calculateSpendingInsights(transactions: Transaction[]): SpendingInsights {
    try {
      const now = new Date();
      const startDate = new Date(now.getTime() - 90 * 24 * 60 * 60 * 1000);
      
      // Group by category
      const byCategory: { [key: string]: { total: number; count: number } } = {};
      
      transactions.forEach(txn => {
        const category = txn.category?.[0] || 'Other';
        if (!byCategory[category]) {
          byCategory[category] = { total: 0, count: 0 };
        }
        byCategory[category].total += txn.amount;
        byCategory[category].count += 1;
      });

      // Sort by total spending
      const sortedCategories = Object.entries(byCategory)
        .map(([name, data]) => ({
          category: name,
          amount: data.total,
          count: data.count,
        }))
        .sort((a, b) => b.amount - a.amount);

      const totalSpending = transactions.reduce((sum, t) => sum + t.amount, 0);
      
      const insights: SpendingInsights = {
        period: {
          start_date: startDate.toISOString().split('T')[0],
          end_date: now.toISOString().split('T')[0],
          days: 90,
        },
        summary: {
          total_spending: totalSpending,
          daily_average: totalSpending / 90,
          monthly_average: totalSpending / 3,
          transaction_count: transactions.length,
        },
        top_categories: sortedCategories.slice(0, 5).map(cat => ({
          category: cat.category,
          amount: cat.amount,
          percentage: (cat.amount / totalSpending) * 100,
          transaction_count: cat.count,
        })),
        recommendations: this.generateRecommendations(byCategory, totalSpending),
      };

      localStorage.setItem(this.INSIGHTS_KEY, JSON.stringify(insights));
      return insights;
    } catch (error) {
      console.error('❌ Error calculating insights:', error);
      return {
        period: { start_date: '', end_date: '', days: 0 },
        summary: { total_spending: 0, daily_average: 0, monthly_average: 0, transaction_count: 0 },
        top_categories: [],
        recommendations: [],
      };
    }
  }

  /**
   * Generate personalized spending recommendations
   */
  private static generateRecommendations(
    byCategory: { [key: string]: { total: number; count: number } },
    totalSpending: number
  ): string[] {
    const recommendations: string[] = [];

    // Check for high food spending
    const foodCategories = Object.entries(byCategory).filter(
      ([name]) => name.toLowerCase().includes('food') || name.toLowerCase().includes('dining')
    );
    const foodTotal = foodCategories.reduce((sum, [_, data]) => sum + data.total, 0);
    if (foodTotal > totalSpending * 0.3) {
      recommendations.push(`💰 You're spending ${(foodTotal / totalSpending * 100).toFixed(1)}% on food. Consider meal planning to save!`);
    }

    // Check for high shopping
    const shoppingTotal = byCategory['Shopping']?.total || 0;
    if (shoppingTotal > 150) {
      recommendations.push(`🛍️ Your shopping spending is high at $${shoppingTotal.toFixed(2)}. Try setting a monthly budget!`);
    }

    // Check for transportation
    const transportTotal = byCategory['Travel']?.total || 0;
    if (transportTotal > 100) {
      recommendations.push(`🚗 Transportation costs are ${(transportTotal / totalSpending * 100).toFixed(1)}% of spending. Consider carpooling!`);
    }

    // General recommendation if none triggered
    if (recommendations.length === 0) {
      recommendations.push('✅ Your spending looks balanced! Keep monitoring your budget.');
    }

    return recommendations;
  }

  /**
   * Get cached insights
   */
  static getCachedInsights(): SpendingInsights | null {
    try {
      const data = localStorage.getItem(this.INSIGHTS_KEY);
      return data ? JSON.parse(data) : null;
    } catch (error) {
      console.error('❌ Error getting cached insights:', error);
      return null;
    }
  }

  /**
   * Clear all Plaid data from localStorage
   */
  static clearAllData(): void {
    try {
      localStorage.removeItem(this.STORAGE_KEY);
      localStorage.removeItem(this.TRANSACTIONS_KEY);
      localStorage.removeItem(this.INSIGHTS_KEY);
      console.log('✅ All Plaid data cleared');
    } catch (error) {
      console.error('❌ Error clearing data:', error);
    }
  }
}
